### ISSUE TITLE
process_video.py script saves empty masks?

### ISSUE THREAD

**Issue by KaareLJensen** [hkchengrex/Cutie#124]
First I just want to complement your project again.
I have now been using it for quite some time and I'm still amazed by Cuties ability to track objects.

By now I have run into an issue I have not so far been able to figure out. And therefore wanted to write here to find out if others have experienced the same or maybe have suggestions how to solve this.

I'm using the `process_video.py` script to process video data.
Per default running using GPU processing.
Now in two different cases, running on two different PC's, the script is always outputting blank mask images. 
And for these two PC setups this happens no matter which data input is used.
Both PC's have been checked to contain valid Nvidia GPU, which are different, but verified to be supported by the PyTorch version used.

But if the scripts device is changed to force the processing to run on CPU, then all is working as expected. With CPU then the processing outputs masks as expected from the input data. 
This make me think that, somehow the GPU is incompatible eventhough it should not be. Or something in the PyTorch CUDA implementation goes wrong for these setups without any error or warnings being shown.

Additionally I have tested the datasets with issues on other PC setup, and here the processing using GPU run just fine returning expected output masks.

If anyone have seen something similar or have suggestions to what might be causing this issue, or maybe have ideas of how to troubleshoot this, I will be happy to hear about it.
Thank you!

**Comment by hkchengrex** [comment]
Thanks a lot for reporting this issue!

Could you try if [scripting_demo.py](https://github.com/hkchengrex/Cutie/blob/main/scripting_demo.py) works? `process_video.py` is provided by a contributor and I am not sure if there are any significant differences that I have missed.

Besides, this reminds me of other issues in similar projects: https://github.com/hkchengrex/XMem/issues/41 and https://github.com/hkchengrex/MiVOS/issues/48 and they were fixed by reinstalling some drivers/libraries. I still do not know what causes these issues.

**Comment by KaareLJensen** [comment]
Hi Rex

I have now tried to run a modified version of the "scripting_demo.py" code.
I need to run it in a package version using PyInstaller and therefore needed to do modifications to exclude the hydra depencency as I cannot get PyInstaller to include this framework. But still this is minor changes.
This is my changed script:

```
import os
from os import chdir, path
import sys
sys.path.append(path.expanduser('cutie'))

from omegaconf import open_dict, OmegaConf

import torch
from torchvision.transforms.functional import to_tensor
from PIL import Image
import numpy as np

from Cutie.cutie.model.cutie import CUTIE
from Cutie.cutie.inference.inference_core import InferenceCore
from Cutie.cutie.inference.utils.args_utils import get_dataset_cfg
from Cutie.cutie.utils.download_models import download_models_if_needed

from omegaconf import open_dict

def get_default_model() -> CUTIE:
    cfg_base = OmegaConf.load('cutie/config/model/base.yaml')
    cfg_video = OmegaConf.load('cutie/config/eval_config.yaml')
    cfg = cfg_video
    cfg.model = cfg_base


    weight_dir = download_models_if_needed()
    with open_dict(cfg):
        cfg['weights'] = os.path.join(weight_dir, 'cutie-base-mega.pth')
    get_dataset_cfg(cfg)

    # Load the network weights
    cutie = CUTIE(cfg).cuda().eval()
    model_weights = torch.load(cfg.weights)
    cutie.load_weights(model_weights)

    return cutie


@torch.inference_mode()
@torch.cuda.amp.autocast()
def test():
    # obtain the Cutie model with default parameters -- skipping hydra configuration
    cutie = get_default_model()
    # Typically, use one InferenceCore per video
    processor = InferenceCore(cutie, cfg=cutie.cfg)
    # the processor matches the shorter edge of the input to this size
    # you might want to experiment with different sizes, -1 keeps the original size
    processor.max_internal_size = 480

    image_path = './examples/images/bike'
    # ordering is important
    images = sorted(os.listdir(image_path))

    # mask for the first frame
    # NOTE: this should be a grayscale mask or a indexed (with/without palette) mask,
    # and definitely NOT a colored RGB image
    # https://pillow.readthedocs.io/en/stable/handbook/concepts.html: mode "L" or "P"
    mask = Image.open('./examples/masks/bike/00000.png')
    assert mask.mode in ['L', 'P']

    # palette is for visualization
    palette = mask.getpalette()

    # the number of objects is determined by counting the unique values in the mask
    # common mistake: if the mask is resized w/ interpolation, there might be new unique values
    objects = np.unique(np.array(mask))
    # background "0" does not count as an object
    objects = objects[objects != 0].tolist()

    mask = torch.from_numpy(np.array(mask)).cuda()

    for ti, image_name in enumerate(images):
        # load the image as RGB; normalization is done within the model
        image = Image.open(os.path.join(image_path, image_name))
        image = to_tensor(image).cuda().float()

        if ti == 0:
            # if mask is passed in, it is memorized
            # if not all objects are specified, we propagate the unspecified objects using memory
            output_prob = processor.step(image, mask, objects=objects)
        else:
            # otherwise, we propagate the mask from memory
            output_prob = processor.step(image)

        # convert output probabilities to an object mask
        mask = processor.output_prob_to_mask(output_prob)

        # visualize prediction
        mask = Image.fromarray(mask.cpu().numpy().astype(np.uint8))
        mask.putpalette(palette)
        mask.show()  # or use mask.save(...) to save it somewhere


if __name__ == '__main__':
    # detect if running from PyInstaller bundle or not 
    if getattr(sys, 'frozen', False) and hasattr(sys, '_MEIPASS'):
        # running in a PyInstaller bundle
        print('running in installer process')
        bundle_cutie_path = sys._MEIPASS + '/cutie' 
        chdir(bundle_cutie_path)
    else:
        # running in a normal Python process
        print('running in a normal Python process')
        chdir(path.expanduser('cutie'))

    test()
```
Still I get the same issue.
Normally the scripting_demo will load the bike images/mask and show one at a time  the four output masks.
But now on this PC with issues, only blank/empty masks are returned.
And still on other PC's the bike masks are returned just fine.
Very anoying and strange.

I tried to look into the other issues you refered to, but have not found any good pointers for me to continue my investigation. They do end with the same case as I see. First output mask is good (but this is also same frame as input mask). All the rest of the output masks are empty.

The PC with issue, is a Windows 11 PC, and does not have any other Python installation, than the packaged version I have installed.
I'm using Python 3.9 with PyTorch 2.2.0 + cu118

**Comment by hkchengrex** [comment]
Thanks for the update.
Unfortunately, I cannot reproduce this problem (nor could I in the two issues that I have linked). 
The "solution" for those two issues was reinstalling the libraries from scratch. I don't know what the mechanism behind that is but I think that is worth a try (I don't see you mentioned it -- sorry if you have already tried this).

For reference, I am using Python 3.11, PyTorch 2.5.1, and CUDA 11.8. But frankly, I have used many different python/pytorch versions to run this project and never had the same issue.

I would love to help more, but this seems difficult since I cannot reproduce the problem. What I would do (if I can reproduce it), is to trace the data from input to output, and see at which step the data misbehave.

**Comment by KaareLJensen** [comment]
Yes this is a tricky issue. 
I'm not able to reproduce the problem on my own PC, as this only happens on a remote PC. Also I'm not allowed to run and test thru a regular Python installation. Therefore having to package the project using PyInstaller.

But you are right, I have not tried to reinstall the libraries from scratch yet. I will do that. And if possible also try to update to use Python 3.11, PyTorch 2.5.1, and CUDA 11.8 as you are using.

It would be nice if I somehow could trace data from input to output to narrow down to see which step is misbehaving.
But as I mentioned I cannot debug directly. Therefore I will need to add logging or saving of intermediated data in the `inference_core.step(...)` method.
I will experiment with this.
As you are the one with the most in-depth understanding of the inner workflow of Cutie, maybe you have some suggestion of which objects in the method will be good to check for mask detection or not and log information about?
(I understand if you do not have time to look into this.)

**Comment by hkchengrex** [comment]
I would start from https://github.com/hkchengrex/Cutie/blob/ec5cdd4cf16f75c73ad785a2f96fb97dbad4125a/cutie/inference/inference_core.py#L172 and go deeper as I narrow the scope.

**Comment by KaareLJensen** [comment]
I have now had the change run some test where I reinstalled the Cutie dependencies and also updated my version of PyTorch.

Reinstallation did not work and still empty mask was returned when running the `scripting_demo.py`.

But using the most recent version of PyTorch v.2.5.1, solved the issue it seems.
I am using Python 3.9 and installing PyTorch and CUDA with
`pip install torch==2.5.1 torchvision==0.20.1 --extra-index-url https://download.pytorch.org/whl/cu118`

Also, for some unknown reason, when setting up my venv for Cutie, I had to reinstall torch last, as some of the Cutie depencencies downgraded my torch version. Just mentioning it here for others refence to double check the correct version being used.

I have not been able to dig deeper into debugging the `inference_core.py` code. So I still have no explaination of why this happens and only for some in very specific PC configurations.

Therefore it seems this can be closed seen from my side.
